Bmi conventions doc (#507)
NOAA-OWP/ngen: doc/BMIconventions.md

# Compliance with the BMI v2.0 Standard

BMI modules written in C, Fortran, and C++ *must* be compiled with the exact same header files as those used with ngen itself, and this means the BMI v2.0 header files provided in the CSDMS [`bmi-c`](https://github.com/csdms/bmi-c/blob/e6f9f8a0ab326218831000b4a5571490ebc21ea2/bmi.h), [`bmi-fortran`](https://github.com/csdms/bmi-fortran/blob/e34cd026f57cabd009ef0f662fc672baee66e442/bmi.f90), and [`bmi-cxx`](https://github.com/csdms/bmi-cxx/blob/be631dc510b477b3bc3eb3c8bbecf3d04ec4005c/bmi.hxx) repositories. Python models must inherit from the `Bmi` class in the [`bmipy`](https://pypi.org/project/bmipy/2.0/) package. Python and C++ classes can be subclasses of the standard BMI interface, but such subclasses must inherit from the standard versions listed here.

> This means among other things that it is not possible to "extend" the BMI interface simply by adding functions to the base interface or header files. In any case, even with subclassed BMI models in Python and C++ with additional functions, ngen would not call any such additional functions because it does not know that they exist. 

Reviewer: "This means, among other things, that it is..." A little comma grammar...

Author: My problem is usually more overuse of commas. This looks like a job for... EMDASH! 🦸‍♂️

> This means among other things that it is not possible to "extend" the BMI interface simply by adding functions to the base interface or header files. In any case, even with subclassed BMI models in Python and C++ with additional functions, ngen would not call any such additional functions because it does not know that they exist. 

Author: ```suggestion
This means&mdash;among other things&mdash;that it is not possible to "extend" the BMI interface simply by adding functions to the base interface or header files. In any case, even with subclassed BMI models in Python and C++ with additional functions, ngen would not call any such additional functions because it does not know that they exist. 
```

   * `get_value(...)` get the model outputs for the just-processed timestep

Other patterns may be allowed by the BMI without being prescriptive, such as passing a whole time series of inputs as an array to `set_value(...)` but this is not supported by ngen. Passing of arrays as input is supported but is applicable in cases where the whole array applies to one timestep (e.g. soil moisture profile or snow layer information for multiple depths).

## `update` vs. `update_until`

For the foreseeable future, ngen will always call `update_until(...)` and will not call a model's `update()` method. Per the [BMI documentation](https://bmi.readthedocs.io/en/stable/#update-until) for `update_until`, an absolute time is passed as the argument, not an incremental time; i.e., if the model start time is `0.0` and ngen is incrementing timesteps by 3600 seconds, the first three iterations will call `update_until` with the values `3600.0`, `7200.0`, and `10800.0`.

At present, ngen will always pass a value greater than the previous value passed to `update_until` to any model instance (it will move only forward in time).

> ## Time Units
> 
> At present, all BMI modules must use seconds as the time units. The output of `get_time_units()` should be `seconds` and the value passed to `update_until(...)` should be interpreted as seconds past time `0` (see below). This is something that we intend to make more flexible in the future.
> 

Reviewer 2: ```suggestion
```
Obviated by #633

Author: Noting that follow-on discussion noted that BMI documentation is very vague about what should be passed to `update_until`, and whether the value returned by `get_time_units()` applies to `update_until()` ... recommend leaving this as is.

* Initialization parameters
* Calibratable parameters
* State inspection (for serialization or debugging)
* Metadata variables (spatial reference information, hyperparameters, etc.) 

These and other uses may be applied by ngen. Specifically, ngen will make use of specified, reserved unpublished variables enumerated later in this document. Therefore, a module should be able to communicate via unpublished variables or at least not malfunction in the case of unpublished variables being used, as described below.

Note that there is currently no discoverability mechanism for unpublished variables, they must be described in documentation (such as this document or a module's  documentation) and often manually implemented or configured for use (such as in a realization config file).

> ## Unknown/unexpected variables passed to `set_value`

Reviewer: Umm...I'm not sure I agree with this statement.  Would be happy to discuss in more detail.

Author: The header? Or the section comments? Sure, here, or hit me up on Slack.

Reviewer: Why should `set_value` not return `BMI_FAILURE`?  If ngen is using unpublished conventional variables, these are essentially OPTIONAL and it *should* explicitly check to ensure that `BMI_SUCCESS` was returned from `set_value` to ensure that it can properly assume the semantics are valid.  `BMI_FAILURE` would indicate that those assumptions are not valid.  That is how calibration parameters are handled right now -- `BMI_FAILURE` just assumes that variable is not calibratable.  When we get to grid conventions, it is almost imperative for `BMI_FAILURE` to be returned if a call to `set_value( 'grid_X_shape')` fails...this allows the model engine to know that the grid cannot be manipulated by the framework, and it can test if the model implements the correct grid functions and their values align in a meaningful fashion.

Author: Ah. Okay, this part I was less sure about. Generally, I wasn't sure how ngen would handle `set_value` returning `BMI_FAILURE` and whether it would consider that an error for a convention variable. You are probably right here, especially if the ngen code already expects these semantics.

The only thing left in my mind is, if the module supports the variable, but encounters an unexpected problem, what then? For C++/Python we can recommend specific exceptions but we're kindof out of options for C/Fortran. Probably no good answer, but it's a likely scenario. Will it be important to distinguish between support+failure vs. nosupport?

Reviewer: Nope, not without enumerated error codes...BMI 3.0 recommendation =)

> Importantly, if an unexpected variable name is passed to `set_value` this *must* not cause an error or instability! Consider the unknown unpublished value to be "offered" and it can be ignored<!-- TODO: " (unless otherwise specified for specific variable names later in this document)"? We might have a required one at some point. -->. Generally, the preferred behavior in the case of an unknown variable is to do nothing and report success (i.e. throw no exception and return `BMI_SUCCESS`).

Author: ```suggestion
Importantly, if an unexpected variable name is passed to `set_value` this *must* not cause an error or instability! Consider the unknown unpublished value to be "offered" and it can be ignored<!-- TODO: " (unless otherwise specified for specific variable names later in this document)"? We might have a required one at some point. -->. Generally, the preferred behavior in the case of an unknown variable is to do nothing (i.e. throw no exception) and return `BMI_FAILURE`&mdash;ngen will interpret this return value as indicating that the variable just set is not supported by the module.
```

## Contiguousness

In order to pass arrays back and forth between modules and ngen (and indrectly from BMI modules to other BMI modules), it is necessary that the arrays being passed are stored in contiguous memory blocks in a known layout. This does not necessarily require that this is how data is stored in memory and used for computation within the model, but when passing data through BMI functions array data must conform to this constraint.

## Zero-Indexing

Arrays should be zero-indexed (the index of the first item is 0, not 1 or some other number) and indexes appearing in the BMI functions (e.g. [`get_value_at_indices`](https://bmi.readthedocs.io/en/stable/#get-value-at-indices)) must be treated as zero-based indices. Of the supported languages, this mainly affects Fortran developers, as Fortran uses 1-based indices by default.

## Layout and Flattening

### C ordering required

The [BMI best practices](https://bmi.readthedocs.io/en/stable/bmi.best_practices.html) document states:

> BMI functions always use flattened, one-dimensional arrays...It’s the developer’s responsibility to ensure that array information is flattened/redimensionalized in the correct order.

> However, strictly speaking, *how* a multi-dimensional array should be flattened into a one-dimensional one is never directly addressed. **For ngen, it is required that flattening happens as if the array was a C array, sometimes referred to as "row major order".** That is, if you create a multi-dimensional array in C (or a C array in C++) it will already be in the appropriate order and layout such that if the data is copied into a 1D array (or the pointer is passed as the result of `get_value_ptr`) it is already in the correct order/layout and will "just work".

Reviewer: It is not guaranteed to be contiguous though if you are using pointers.  Should probably note that caveat.

Author: Ah, had to reread... okay, yes, if you create a C array of pointers that's different. Is there a name for the "correct" allocation method?

> However, strictly speaking, *how* a multi-dimensional array should be flattened into a one-dimensional one is never directly addressed. **For ngen, it is required that flattening happens as if the array was a C array, sometimes referred to as "row major order".** That is, if you create a multi-dimensional array in C (or a C array in C++) it will already be in the appropriate order and layout such that if the data is copied into a 1D array (or the pointer is passed as the result of `get_value_ptr`) it is already in the correct order/layout and will "just work".

Author: ```suggestion
However, strictly speaking, *how* a multi-dimensional array should be flattened into a one-dimensional one is never directly addressed. **For ngen, it is required that flattening happens as if the array was a contiguous C array, sometimes referred to as "row major order".** That is, if you create a contiguous (i.e. not using pointers) multi-dimensional array in C (or a C array in C++) it will already be in the appropriate order and layout such that if the data is copied into a 1D array (or the pointer is passed as the result of `get_value_ptr`) it is already in the correct order/layout and will "just work".
```

(See also https://www.visitusers.org/index.php?title=C_vs_Fortran_memory_order )

C/C++
```C
float var[2][3][4];
int x, y, z;
float v = 0.0;
for(z = 0; z < 2; z++)
    for(y = 0; y < 3; y++)
        for(x = 0; x < 4; x++)
            var[z][y][x] = v += 0.01;
```
<!--
float var_flat[24];
memcpy(var_flat, var, 24*sizeof(float));
printf("%f %f %f %f %f\n", var[0][0][0], var[0][0][1], var[0][0][2], var[0][1][0], var[1][0][0] );
printf("%f %f %f %f %f\n", var_flat[0],  var_flat[1],  var_flat[2],  var_flat[4],  var_flat[12] );
-->

Fortran:
<!--
real, dimension(0:1,0:2,0:3) :: var
integer:: x,y,z
real:: v = 0.0
zloop: do z = 0, 1
   yloop: do y = 0, 2
      xloop: do x = 0, 3
        v = v + 0.01
        var(z,y,x) = v ! Note the indices ordering here
      end do xloop
   end do yloop  
end do zloop
-->
```Fortran
real, dimension(0:3,0:2,0:1) :: var ! Note the reversal of the dimension sizes
integer:: x,y,z
real:: v = 0.0
zloop: do z = 0, 1
    yloop: do y = 0, 2
        xloop: do x = 0, 3
            v = v + 0.01

>             var(x,y,z) = v ! Note the indices ordering here

Reviewer 2: `3, 2, 1` != `4, 3, 2` like I'd expect for equivalence to the other examples here

Author: Wait, what? It's been a while but I ended up running these examples a lot... you saw the `0:...` up in the `dimension(...)` call, right?

Reviewer 2: Ah, I see now. I forgot that the way Fortran writes things ends up with closed intervals, not half-open intervals like C or Python.

### When this matters

#### Fortran models using z,y,x index ordering

If you have a Fortran model and are using `(y,x)` or `(z,y,x)` ordering for your array indices, then the flattened, linear contiguous memory representation of your array is not the same as C ordering. In this case, the BMI interface layer in your module will have to copy/reorganize the memory in the array to be in C order whenever responding to BMI functions--essentially, recopy `(z,y,x)` arrays to be in `(x,y,z)` order. This is necessary when exposing an array outside of your model module because other models and ngen will assume that the `get_value_at_index` and `set_value_at_index` operations work a certain way on the memory representation and that the array is in a specific layout when it represents spatial data. See below for details on both of these scenarios.

#### Clarification for `get_value_at_indices` and `set_value_at_indices`

The BMI documentation for [`get_value_at_indices`]() and [`set_value_at_indices`]() states that the `inds` parameter designates:

> ...the locations specified by the one-dimensional array indices in the `inds` argument...Both `dest` and `inds` are flattened arrays....

> However if read perfectly literally, `inds` is by necessity already a one-dimensional array and therefore can't be flattened--rather, the implication is that the indexes in `inds` are *indexes into a flattened representation of the target variable array*. That is, in the example above, an `inds` array containing `[2, 21]` should return `[0.03, 0.22]`. However, if the flatteing is assumed to be done differently by different parts of the system, the wrong values may be retrieved or set.

Reviewer 2: ```suggestion
However if read perfectly literally, `inds` is by necessity already a one-dimensional array and therefore can't be flattened--rather, the implication is that the indexes in `inds` are *indexes into a flattened representation of the target variable array*. That is, in the example above, an `inds` array containing `[2, 21]` should return `[0.03, 0.22]`. However, if the flattening is assumed to be done differently by different parts of the system, the wrong values may be retrieved or set.
```
